arch/lkl/kernel/signal.c

```diff
@@ -28,47 +31,249 @@ static void initialize_uctx(struct ucontext *uctx, const struct pt_regs *regs)
             break;
         }
     }
 }
 
-void do_signal(struct pt_regs *regs)
+/*
+ * Get the end of the list and make it point at the new node,
+ * if there were no nodes before then bot head and tail point at this one node
+ */
+
+void append_ksignal_node(struct thread_info *task, struct ksignal *ksig)
+{
+    struct ksignal_list_node *ptr;
+    
+    /* first make a place to hold our signal */
+    struct ksignal_list_node *node = kmalloc(sizeof(struct ksignal_list_node), GFP_ATOMIC);     
+    if (node == NULL) {
+        lkl_bug("Unable to allocate memory to hold signal state.");
+    }
+    memcpy(&node->sig, ksig, sizeof(*ksig));
+    node->next = NULL;
+
+    /* now put it in the linked list */
+    spin_lock(&task->signal_list_lock);
+    ptr = task->signal_list_tail;
+    if (ptr == NULL) { /* list is empty so point both head and tail at the given node */
+        task->signal_list_head = node;
+        task->signal_list_tail = node;
+    } else { /* otherwise just the tail changes */
+        ptr->next = node; 
+    }
+
+    spin_unlock(&task->signal_list_lock);
+}    
+
+/*
+ * Must be called with the cpu lock held
+ */
+
+void move_signals_to_task(void)
+{
+    struct ksignal ksig;    /* place to hold retrieved signal */
+    /*
+     * Get the lkl local version of the current task, so we can store the signals
+     * in a list hanging off it.
+     */
+    struct thread_info *current_task;
+
+    /* In DEBUG build check that the lock is currently held by this task */
+    lkl_assert_cpu_owner();
+
+    current_task = task_thread_info(current); 
+    /* note that get_signal may never return if the task is dead (eg pending a SIGKILL) */
+    while (get_signal(&ksig)) {	
+        LKL_TRACE("Appending task %p signal %d\n", current, ksig.sig);
+        append_ksignal_node(current_task, &ksig);
+
+/* 
+ * TODO check whether the task is runnable here. If it is INTERUPTABLE then set it to RUNNING
+ * so it can be scheduled. Check with the ltp test tgkill01.
+ * 
+ * This will be another PR.
+ */
+
+    }
+}
+
+/*
+ * Must be called WITH the cpu lock held.
+ * The signal invoke code drops and retakes the lock.
+ */
+
+void send_current_signals(struct ucontext *uctx)
 {
     struct ksignal ksig;
     struct ucontext uc;
-
+    
     LKL_TRACE("enter\n");
 
-    memset(&uc, 0, sizeof(uc));
-    initialize_uctx(&uc, regs);
-    while (get_signal(&ksig)) {
+    if (uctx == NULL) {
+        uctx = &uc;
+        memset(uctx, 0, sizeof(uc));
+        initialize_uctx(uctx, NULL);
+    }
+
+    /* In DEBUG build check that the lock is currently held by this task */
+    lkl_assert_cpu_owner();
+
+    /* get pending signals */
+    while (get_next_ksignal(task_thread_info(current), &ksig)) {
         LKL_TRACE("ksig.sig=%d", ksig.sig);
+        
+        /* Actually deliver the signal. */
 
-        /* Whee!  Actually deliver the signal.  */
-        handle_signal(&ksig, &uc);
+        handle_signal(&ksig, uctx); 
```

Comment:
Is there a return value? If so do we care about the return value here?

Reply:
handle_signal is void. I don't believe posix style signals can directly return anything to the calling code.

Reply:
Yes, I think you are right about posix signals not being able to return anything meaningful.